scripts/build/check_init_priorities.py

```diff
@@ -0,0 +1,316 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Google LLC
+# SPDX-License-Identifier: Apache-2.0
+
+"""
+Checks the initialization priorities
+
+This script parses the object files in the specified build directory, creates a
+list of known devices and their effective initialization priorities and
+compares that with the device dependencies inferred from the devicetree
+hierarchy.
+
+This can be used to detect devices that are initialized in the incorrect order,
+but also devices that are initialized at the same priority but depends on each
+other, which can potentially break if the linking order is changed.
+"""
+
+import argparse
+import logging
+import os
+import pathlib
+import pickle
+import sys
+
+from elftools.elf.elffile import ELFFile
+from elftools.elf.relocation import RelocationSection
+from elftools.elf.sections import SymbolTableSection
+
+# This is needed to load edt.pickle files.
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..",
+                                "dts", "python-devicetree", "src"))
+from devicetree import edtlib  # pylint: disable=unused-import
```

Comment:
@keith-zephyr and I discovered a bug where if you do this import, but use the EDT pickle, the module may not be the same between objects from the pickle and from edtlib.  e.g., you may not be able to have `except edtlib.EDTError`, as the underlying exception type objects will differ in ID.

Not sure if this affects you here but it's something to keep in mind.  Our workaround was to use `inspect.getmodule(edt_pickle)` to get the underlying module from the loaded pickle instead of importing it directly.

(Side note, pickles are usually extremely sketchy and this is why stuff like this comes out to bite!  We probably shouldn't be using them at all, but I know this pattern already existed here)

Reply:
I see, I'm not doing anything with the object so I think I should be ok for this one.